kb/0003-style-guide.md

```diff
@@ -0,0 +1,1004 @@
+
+Stroustrup, the original designer of C++, [makes this point](http://www.stroustrup.com/bs_faq2.html#exceptions-why) much better than I ever could.
+
+## Use C++-style cast instead of C-style cast
+Use the C++-style cast (static\_cast<>, dynamic\_cast<> ...) instead of the C-style cast. The C++-style cast allows more compiler checks and is considerably safer.
+
+```cpp
+// Bad Idea
+double x = getX();
+int i = (int) x;
+
+// Not a Bad Idea
+int i = static_cast<int>(x);
+```
+Additionally the C++ cast style is more visible and has the possibility to search for.
+
+But consider refactoring of program logic (for example, additional checking on overflow and underflow) if you need to cast `double` to `int`. Measure three times and cut 0.9999999999981 times.
+
+## Do not define a variadic function
+Variadic functions can accept a variable number of parameters. The probably best known example is printf(). You have the possibility to define this kind of functions by yourself but this is a possible security risk. The usage of variadic functions is not type safe and the wrong input parameters can cause a program termination with an undefined behavior. This undefined behavior can be exploited to a security problem.
+If you have the possibility to use a compiler that supports C++11, you can use variadic templates instead.
+
+[It is technically possible to make typesafe C-style variadic functions with some compilers](https://github.com/lefticus/cppbestpractices/issues/53)
+
+## Additional Resources
+
+[How to Prevent The Next Heartbleed](http://www.dwheeler.com/essays/heartbleed.html) by David Wheeler is a good analysis of the current state of code safety and how to ensure safe code.
+
+# Considering Maintainability
+
+
+## Avoid Compiler Macros
+
+Compiler definitions and macros are replaced by the preprocessor before the compiler is ever run. This can make debugging very difficult because the debugger doesn't know where the source came from.
+
+```cpp
+// Bad Idea
+#define PI 3.14159;
+
+// Good Idea
+namespace storm
+{
+class Constants
+{
+  public:
+    // if the above macro would be expanded, then the following line would be:
+    //   static const double 3.14159 = 3.14159;
+    // which leads to a compile-time error. Sometimes such errors are hard to understand.
+    static constexpr double PI = 3.14159;
+};
+}
+```
+
+## Consider Avoiding Boolean Parameters
+
+They do not provide any additional meaning while reading the code. You can either create a separate function that has a more meaningful name, or pass an enumeration that makes the meaning more clear.
+
+See http://mortoray.com/2015/06/15/get-rid-of-those-boolean-function-parameters/ for more information.
+
+## Avoid Raw Loops
```

Comment:
This is a controversial point. In general, simple loops are way more expressive. especially range-based loops